include/ginkgo/core/config/property_tree.hpp

```diff
@@ -0,0 +1,181 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_CONFIG_PROPERTY_TREE_HPP_
+#define GKO_PUBLIC_CORE_CONFIG_PROPERTY_TREE_HPP_
+
+
+#include <cassert>
+#include <exception>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+
+#include <ginkgo/core/config/data.hpp>
+
+
+namespace gko {
+namespace config {
+
+
+/**
+ * pnode is to describe the property tree
+ */
+class pnode {
+public:
+    using key_type = std::string;
+    using data_type = data;
+
+    enum class status_t { empty, array, data, map };
+
+    pnode() : status_(status_t::empty) {}
+
+    pnode(const data_type& d) : status_(status_t::data), data_(d) {}
+
+    pnode(const std::vector<pnode>& array)
+        : status_(status_t::array), array_(array)
+    {}
+
+    pnode(const std::map<key_type, pnode>& list)
+        : status_(status_t::map), list_(list)
+    {}
+
+    // bool conversion. It's true if and only if it contains data.
+    operator bool() const noexcept { return status_ != status_t::empty; }
+
+    const data_type& get_data() const
+    {
+        this->throw_if_not_contain(status_t::data);
+        return data_;
+    }
+
+    std::vector<pnode>& get_array()
```

Comment:
It might still be necessary later on if we want to allow building configs from existing objects.

Reply:
That goes quite a bit further than the scope that is discussed here. And even then I don't think it would be strictly necessary.

Reply:
Yes, it would probably be sufficient to provide an `vector<pnode>` constructor. Just wanted to mention it, because I think it will be useful for people to quickly move to the config model from the factory model.